### PR TITLE
Add function-body to SoftmaxGrad

### PR DESCRIPTION
**Pull request by gramalingam**
**Description**:
(a) Add function-body to the SoftmaxGrad schema.
(b) Add test-cases to validate the function-body by running the op and the expanded function-body using the CPU provider to establish equivalence.
(c) Add type-context when calling the function-body-builder (which will be required for other ops for typed constants).

**Motivation and Context**
This will enable, for example, translation of ONNX graphs containing calls to SoftmaxGrad to ONNX-MLIR using the function mechanism.

### CODE REVIEW

File: onnxruntime/core/graph/graph.cc [hunk 1]
```diff
@@ -2371,7 +2371,16 @@ void Graph::InitFunctionBodyForNode(Node& node) {
...
-      onnx::FunctionBodyBuildContextImpl function_body_ctx(node_proto);
+      std::vector<TypeProto> input_types;
+      for (size_t i = 0, n = node.InputDefs().size(); i < n; i++) {
+        auto p_node_arg = node.InputDefs().at(i);
+        if ((nullptr != p_node_arg) && p_node_arg->Exists()) {
+          auto& type = *(p_node_arg->TypeAsProto());
+          input_types.emplace_back(type);
+        } else
+          input_types.emplace_back();
+      }
```
Comment (KeDengMS) [inline onnxruntime/core/graph/graph.cc:2382]: May use node.ForEachDef

File: onnxruntime/core/graph/contrib_ops/onnx_function_util.cc [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    for (const auto& inp : node.inputs) {
+      np->add_input(inp);
+    }
+    for (const auto& o : node.outputs) {
+      np->add_output(o);
+    }
+    for (const auto& attr : node.attributes) {
+      *(np->add_attribute()) = attr.proto;
+    }
+  }
```
Comment (KeDengMS) [inline onnxruntime/core/graph/contrib_ops/onnx_function_util.cc:41]: np->CopyFrom()?

File: orttraining/orttraining/test/gradient/function_ops_test.cc [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+typedef std::vector<onnxruntime::NodeArg*> ArgMap;
+
+static ONNX_NAMESPACE::TypeProto TensorType(int32_t elem_type, std::vector<int64_t> dims) {
+  ONNX_NAMESPACE::TypeProto typeProto;
+  typeProto.mutable_tensor_type()->set_elem_type(elem_type);
+  auto* shape = typeProto.mutable_tensor_type()->mutable_shape();
+  for (auto dim : dims)
+    shape->add_dim()->set_dim_value(dim);
+  return typeProto;
+}
```
Comment (KeDengMS) [inline orttraining/orttraining/test/gradient/function_ops_test.cc:31]: Could we add test case like [batch,seq,512]?

Comment (gramalingam) [reply orttraining/orttraining/test/gradient/function_ops_test.cc:31]: There is already a 3D test-case. But it has default axis. Do you want a 3D-case with some other axis-value?

File: onnxruntime/core/graph/graph.cc (hunk 1 above)
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc:2382]: We don't want to skip the non-exist cases here: we want to push a default-value for these cases. So, I am not sure we can do that.

File: onnxruntime/core/graph/contrib_ops/onnx_function_util.cc (hunk 2 above)
Comment (gramalingam) [reply onnxruntime/core/graph/contrib_ops/onnx_function_util.cc:41]: The source is a NodeDef, not a NodeProto.

File: orttraining/orttraining/test/gradient/function_ops_test.cc (hunk 3 above)
Comment (KeDengMS) [reply orttraining/orttraining/test/gradient/function_ops_test.cc:31]: I was thinking about symbolic dim vs. static dim. Seems the test case only handles static dims.

---
In reply to: [592862644](https://github.com/microsoft/onnxruntime/pull/6988#discussion_r592862644) [](ancestors = 592862644)

**Comment by souptc** [comment]
LGTM

File: orttraining/orttraining/test/gradient/function_ops_test.cc [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+  testCase.opsets[kMSDomain] = 1;
+  InitSoftmaxGradTestCase(testCase, {3, 2, 2});
+
+  auto model1 = testCase.CreateModel();
+  auto results1 = onnxruntime::test::Run(*model1, testCase.input_value_map, testCase.output_names);
+
+  testCase.opsets[kOnnxDomain] = 13;
+  testCase.opsets[kMSDomain] = 1;
+
+  auto model2 = testCase.CreateModel();
```
Comment (souptc) [inline orttraining/orttraining/test/gradient/function_ops_test.cc:269]: sign off

**Review by souptc — APPROVED** [review]

**Pull request merged**